### PR TITLE
[SQL VM] `az sql vm update`: Add configuration options for SQL Assessment pre-requisites

### PR DESCRIPTION
**Pull request by vadeveka**
**Related command**
`az sql vm update --enable-assessment true -g <Resource Group Name> -n <SQL VM Name> --workspace-name <LA workspace's name> --workspace-rg <LA workspace's resource group name>`

**Description**
- SQL VM offers ability to configure and run assessment via [previous PR](https://github.com/Azure/azure-cli/pull/21281). This PR adds support for verifying and managing pre-requisites such as Log Analytics extension on VM and custom log definition on workspace.
   - Re-use `az vm`'s code for managing Log Analytics extension. See `src/azure-cli/azure/cli/command_modules/sqlvm/_util.py`
   - Rename `SQL Best Practice Assessment` to `SQL best practices assessment` as per branding requirement
   - Move LogAnalyticsWorkspacePreparer to cli-test-sdk for use across modules
- Fix usage of is_r_services field when populating request object

**Testing Guide**
```
# Deploy pre-requisites for SQL best practices assessment using given LA workspace
az sql vm update --enable-assessment true -g <Resource Group Name> -n <SQL VM Name> --workspace-name <LA workspace's name> --workspace-rg <LA workspace's resource group name>

# Deploy pre-requisites for SQL best practices assessment using workspace already associated with VM, if not throw appropriate error
az sql vm update –assessment-weekly-interval 3 --assessment-day-of-week Sunday –assessment-start-time-local "23:00" -g <Resource Group Name> -n <SQL VM Name>
```
Scenarios:
1. No workspace attached to VM. User does not provide workspace arguments -> Throw error
2. No workspace attached to VM. User provides workspace arguments -> Deploy VM extension and custom log if not present.
3. Workspace attached to VM. User does not provide workspace arguments -> Determine workspace from extension. Verify workspace exists. Deploy custom log if necessary.
4. Workspace attached to VM. User provides workspace arguments -> Determine workspace from extension. Verify workspace exists and matches user arguments. Deploy custom log if necessary.
5. Configure assessment/workspace via CLI on 1 VM. Configure via Portal for another VM for same workspace -> Custom log definition created by CLI also recognized by Portal

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
SQL VM

File: src/azure-cli-testsdk/azure/cli/testsdk/preparers.py [hunk 1]
```diff
@@ -465,6 +467,37 @@ def _get_virtual_network(self, **kwargs):
                        'decorator @{} in front of this VirtualNetworkNic preparer.'
             raise CliTestError(template.format(VnetNicPreparer.__name__))
 
+
+class LogAnalyticsWorkspacePreparer(AbstractPreparer, SingleValueReplacer):
```
Comment (evelyn-ys) [inline src/azure-cli-testsdk/azure/cli/testsdk/preparers.py:471]: Please keep align with other preparers, others inherits from `NoTrafficRecordingPreparer, SingleValueReplacer`

File: src/azure-cli-testsdk/azure/cli/testsdk/preparers.py [hunk 2]
```diff
@@ -465,6 +467,37 @@ def _get_virtual_network(self, **kwargs):
...
+        super(LogAnalyticsWorkspacePreparer, self).__init__(name_prefix, 15)
+        self.location = location
+        self.parameter_name = parameter_name
+        self.resource_group_parameter_name = resource_group_parameter_name
+        self.skip_delete = skip_delete
+
+    def create_resource(self, name, **kwargs):
+        group = self._get_resource_group(**kwargs)
+        template = ('az monitor log-analytics workspace create -l {} -g {} -n {}')
+        execute(DummyCli(), template.format(self.location, group, name))
```
Comment (evelyn-ys) [inline src/azure-cli-testsdk/azure/cli/testsdk/preparers.py:483]: Please use `live_only_execute()` like other preparers

File: src/azure-cli-testsdk/azure/cli/testsdk/preparers.py [hunk 3]
```diff
@@ -465,6 +467,37 @@ def _get_virtual_network(self, **kwargs):
...
+        return {self.parameter_name: name}
+
+    def remove_resource(self, name, **kwargs):
+        if not self.skip_delete:
+            group = self._get_resource_group(**kwargs)
+            template = ('az monitor log-analytics workspace delete -g {} -n {} --yes')
+            execute(DummyCli(), template.format(group, name))
```
Comment (evelyn-ys) [inline src/azure-cli-testsdk/azure/cli/testsdk/preparers.py:490]: same here

File: src/azure-cli/azure/cli/command_modules/sqlvm/_util.py [hunk 4]
```diff
@@ -3,6 +3,20 @@
...
+from azure.core.paging import ItemPaged
+from azure.cli.command_modules.vm._client_factory import cf_log_analytics_data_sources
+from azure.cli.core.commands.client_factory import get_subscription_id
+from azure.cli.command_modules.vm.custom import (
+    list_extensions,
+    _get_log_analytics_client,
+    set_extension,
+    extension_mappings,
+    _WINDOWS_OMS_AGENT_EXT
+)
```
Comment (evelyn-ys) [inline src/azure-cli/azure/cli/command_modules/sqlvm/_util.py:16]: Add VM owner @zhoxing-ms for review.

Do you agree that `sqlvm` binding with `vm` functions?

Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/sqlvm/_util.py:16]: I think it is OK to call the public method from `vm` module. 
For calling private methods, this way does not conform to the code specification, and it is easy to ignore the calling relationship. So I personally suggest that either implement `_get_log_analytics_client()` method in `sqlvm` module *(the code of the `_get_log_analytics_client()` method is very simple)* or change the `_get_log_analytics_client()` to a public method

Comment (vadeveka) [reply src/azure-cli/azure/cli/command_modules/sqlvm/_util.py]: Makes sense. Copied all private variables to sqlvm as it was minimal code.

**Review by zhoxing-ms — APPROVED** [review]

**Review by evelyn-ys — APPROVED** [review]

**Review by kairu-ms — APPROVED** [review]

**Pull request merged**